### PR TITLE
MAINT: Replace append_metastr_to_string function.

### PR DESCRIPTION
**Pull request by charris**
This is part of removing the PyUString_ConcatAndDel cleanup and allows
 us to write more readable code when generating strings. There is also 
more checking for errors after this change.

### CODE REVIEW

File: numpy/core/src/multiarray/descriptor.c [hunk 1]
```diff
@@ -1892,18 +1892,28 @@ arraydescr_protocol_typestr_get(PyArray_Descr *self)
...
+        PyObject *umeta = metastr_to_unicode(meta, 0);
+        if (umeta == NULL) {
+            Py_DECREF(ret);
+            return NULL;
+        }
 
-        ret = append_metastr_to_string(meta, 0, ret);
+        PyObject *tmp = ret;
+        ret = PyUnicode_Concat(umeta, ret);
+        Py_DECREF(tmp);
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/descriptor.c:1914]: Could if you wanted use
```suggestion
        Py_SETREF(ret, PyUnicode_Concat(umeta, ret));
```

File: numpy/core/src/multiarray/scalartypes.c.src [hunk 2]
```diff
@@ -554,32 +548,34 @@ timedeltatype_repr(PyObject *self)
 
     /* The value */
     if (scal->obval == NPY_DATETIME_NAT) {
-        ret = PyUnicode_FromString("numpy.timedelta64('NaT'");
+        val = PyUnicode_FromString("'NaT'");
```
Comment (eric-wieser) [inline numpy/core/src/multiarray/scalartypes.c.src:551]: This is a (good) behavior change for how `np.timedelta('NaT', 'D')` displays, right?

**Review by eric-wieser — APPROVED** [review]
Generally looks good - I have a nasty feeling I started on these changes myself and never pushed them...

File: numpy/core/src/multiarray/scalartypes.c.src (hunk 2 above)
Comment (charris) [reply numpy/core/src/multiarray/scalartypes.c.src:551]: Shouldn't change, looks like
```
In [3]: np.timedelta64('NaT', 'D')                                              
Out[3]: numpy.timedelta64('NaT','D')
```
How would you like it to display?

Comment (eric-wieser) [reply numpy/core/src/multiarray/scalartypes.c.src:551]: My mistake, I misread the old code as `numpy.timedelta64('NaT')`

**Comment by charris** [comment]
Oops, looks like I needed to test more than core

**Comment by eric-wieser** [comment]
Thanks, glad this made it in, and sorry for not pushing whatever local work I had!

**Comment by charris** [comment]
How much local work do you have? I going forward with other replacements.

**Comment by eric-wieser** [comment]
I'll check now

**Comment by eric-wieser** [comment]
I think this was the only one that was in a mergeable state - and your patch is slightly wider-reaching than mine was :)

**Comment by charris** [comment]
Darn :) Well, back to the slave pits...

**Pull request merged**